File: src/integrationTest/java/com/naminhyeok/fantazzk/template/TemplateRepositoryIntegrationTest.java
package com.naminhyeok.fantazzk.template;

import static org.assertj.core.api.Assertions.assertThat;

import jakarta.persistence.EntityManager;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestConstructor;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest(
    properties = {
        "spring.datasource.url=jdbc:h2:mem:template-repository-test;DB_CLOSE_DELAY=-1",
        "spring.datasource.driver-class-name=org.h2.Driver",
        "spring.datasource.username=sa",
        "spring.datasource.password=",
        "spring.jpa.hibernate.ddl-auto=create-drop",
        "spring.liquibase.enabled=false",
        "sentry.enabled=false"
    }
)
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
@RequiredArgsConstructor
class TemplateRepositoryIntegrationTest {
    private final Templates templates;
    private final EntityManager entityManager;

    @Test
    @Transactional
    void 템플릿과_선수_컬렉션을_저장하고_다시_읽는다() {
        Template saved =
            templates.save(
                Template.createAuction(
                    "주말 풋살 경매전",
                    2,
                    2,
                    300,
                    List.of("선수1", "선수2")
                )
            );

        entityManager.flush();
        entityManager.clear();

        Template reloaded = templates.findById(saved.getId()).orElseThrow();

        assertThat(reloaded.getId()).isEqualTo(saved.getId());
        assertThat(reloaded.getName()).isEqualTo("주말 풋살 경매전");
        assertThat(reloaded.getMode()).isEqualTo(TemplateMode.AUCTION);
        assertThat(reloaded.getPlayers().stream().map(TemplatePlayer::getName))
            .containsExactly("선수1", "선수2");
    }
}

File: team-building/repository-jdbc/src/integrationTest/kotlin/com/naminhyeok/fantazzk/teambuilding/repository/TemplateRepositoryIntegrationTest.kt
package com.naminhyeok.fantazzk.teambuilding.repository

import com.naminhyeok.fantazzk.teambuilding.TeamBuildingMode
import com.naminhyeok.fantazzk.teambuilding.config.TeamBuildingJdbcConfiguration
import com.naminhyeok.fantazzk.teambuilding.template.Template
import com.naminhyeok.fantazzk.teambuilding.template.TemplateIdentity
import com.naminhyeok.fantazzk.teambuilding.template.TemplatePlayer
import com.naminhyeok.fantazzk.teambuilding.template.of
import com.naminhyeok.fantazzk.teambuilding.template.repository.TemplatePlayerRepository
import com.naminhyeok.fantazzk.teambuilding.template.repository.TemplateRepository
import com.naminhyeok.fantazzk.teambuilding.template.repository.TemplateRepositoryAutoConfiguration
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.autoconfigure.ImportAutoConfiguration
import org.springframework.boot.data.jdbc.test.autoconfigure.DataJdbcTest
import org.springframework.boot.jdbc.test.autoconfigure.AutoConfigureTestDatabase
import org.springframework.boot.liquibase.autoconfigure.LiquibaseAutoConfiguration
import org.springframework.test.context.TestConstructor

@ImportAutoConfiguration(
    LiquibaseAutoConfiguration::class,
    TeamBuildingJdbcConfiguration::class,
    TemplateRepositoryAutoConfiguration::class,
)
@DataJdbcTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
class TemplateRepositoryIntegrationTest(
    private val cut: TemplateRepository,
    private val templatePlayerRepository: TemplatePlayerRepository,
) {
    @Test
    fun `템플릿을 저장하고 조회할 수 있다`() {
        val saved =
            cut.save(
                Template(
                    name = "테스트 경매",
                    mode = TeamBuildingMode.AUCTION,
                    teamCount = 5,
                    teamSize = 5,
                    budget = 300,
                ),
            )

        assertThat(saved.templateId).isGreaterThan(0)

        val found = cut.findById(TemplateIdentity.of(saved.templateId))
        assertThat(found).isNotNull
        assertThat(found!!.mode).isEqualTo(TeamBuildingMode.AUCTION)
        assertThat(found.budget).isEqualTo(300)
    }

    @Test
    fun `템플릿 선수를 저장하고 조회할 수 있다`() {
        val template =
            cut.save(
                Template(name = "드래프트", mode = TeamBuildingMode.DRAFT, teamCount = 2, teamSize = 3),
            )

        templatePlayerRepository.saveAll(
            listOf(
                TemplatePlayer(templateId = template.templateId, name = "선수1", displayOrder = 0),
                TemplatePlayer(templateId = template.templateId, name = "선수2", displayOrder = 1),
            ),
        )

        val found = templatePlayerRepository.findByTemplateId(template.templateId)
        assertThat(found).hasSize(2)
        assertThat(found.map { it.name }).containsExactly("선수1", "선수2")
    }
}
